Custom-Pylint-checks.md

```diff
@@ -0,0 +1,569 @@
+
+**If you want to prevent code that behaves a certain way, any linter you write must miss some code that performs the undesired behavior, raise an error on benign code, or both.**
+
+## Write checkers
+
+### Types of checkers
+
+Pylint supports three kinds of checkers:
+
+* **AST checkers** operate on an [abstract syntax tree (AST)](https://en.wikipedia.org/wiki/Abstract_syntax_tree) of the code that Pylint generates. The checker provides functions to be called as Pylint traverses the AST, and these functions can raise lint errors.
+
+* **Token checkers** operate on a list of the tokens that represents the code being linted. The tokens are of the same form as the tokens generated by [`tokenize.tokenize()`](https://docs.python.org/3/library/tokenize.html#tokenize.tokenize). To get familiar with how files get tokenized, put some simple Python code into `test.py` and then run:
+
+  ```console
+  python -m tokenize test.py
+  ```
+
+  For example, consider this `test.py`:
+
+  ```python
+  def test(arg):
+      print(arg)
+      return arg
+  ```
+
+  Here is the token representation of the file:
+
+  ```console
+  $ python -m tokenize test.py
+  0,0-0,0:            ENCODING       'utf-8'
+  1,0-1,3:            NAME           'def'
+  1,4-1,8:            NAME           'test'
+  1,8-1,9:            OP             '('
+  1,9-1,12:           NAME           'arg'
+  1,12-1,13:          OP             ')'
+  1,13-1,14:          OP             ':'
+  1,14-1,15:          NEWLINE        '\n'
+  2,0-2,4:            INDENT         '    '
+  2,4-2,9:            NAME           'print'
+  2,9-2,10:           OP             '('
+  2,10-2,13:          NAME           'arg'
+  2,13-2,14:          OP             ')'
+  2,14-2,15:          NEWLINE        '\n'
+  3,4-3,10:           NAME           'return'
+  3,11-3,14:          NAME           'arg'
+  3,14-3,15:          NEWLINE        '\n'
+  4,0-4,0:            DEDENT         ''
+  4,0-4,0:            ENDMARKER      ''
+  ```
+
+* **Raw checkers** operate on the raw code without any pre-processing by Pylint. You should avoid writing custom raw checkers since they are less efficient than token or AST checkers.
+
+As we work through the example below, we'll cover each of the three kinds of checkers.
+
+### Hello world
+
+To get a better idea for how lint checks work, let's try writing a simple Pylint rule that forbids the string `Hello, world!`. (Maybe you don't want people accidentally opening PRs that include code that prints `Hello, world!` from when they were learning Python.) First consider what kinds of code should violate the rule. Here are some examples:
+
+```python
+s = "Hello, world!";
+
+print("Hello, world!");
+```
+
+Below, we'll walk through how to implement this rule using each of the three kinds of checkers in Pylint: AST checkers, token checkers, and raw checkers.
```

Comment:
I'm a bit confused about the process that I should follow for writing a new lint check. How would I know which type of checker I should use for this case? What are the guidelines for when I should use which type of checker?

Also, do I commit to a type of checker and only then consider the pros/cons in terms of false positives/negatives? The structure of this doc suggests that this is the case but that feels a bit backwards; it might be worth providing a bit more explanation here in terms of how a developer should reason about which path to take.

Reply:
done